### PR TITLE
Corrected is_tangent in ellipse.py

### PR DESCRIPTION
**Pull request by Amit-Modi**
What's Wrong: #19471
```python
>>> from sympy import *
>>> c = Circle(Point(0,0),2)
>>> poly1 = Polygon((-2,-2),(-2,2),(2,2),(3,0),(2,-2))
>>> c.is_tangent(poly1)
True
>>> poly2 = Polygon((2,2),(3,0),(2,-2),(-2,-2),(-2,2))
>>> c.is_tangent(poly2)
False
>>> r = Ray(Point(0,2),Point(2,2))
>>> c.is_tangent(r)
False
```

notice that poly1 and poly2 are giving different results,
but are same polygon except there starting points are different.
Further, in my opinion( I found very less on this topic on internet but by definition in google),  tangent is a straight line or plane that touches a curve or curved surface at a point, but if EXTENDED does not cross it at that point.
the ray,r touches the circle,c at r.source but when extended does not cross circle,c at any point. by this definition r is a tangent to circle c.

Changes I made:
change 1-
Line Number:720-730
       if one or more sides of polygon are tangent to ellipse and other does not touch it, polygon is a tangent
      

change 2-
Line Number: 714
       original content : 
                return intersect[0] != o.source and not self.encloses_point(o.source)
       my change:
                return o in self.tangent_lines(intersect[0])[0]

### CODE REVIEW

**Comment by certik** [comment]
Hi @Amit-Modi, thanks for the fix. Can you please also write a test, that failed before, but now it works? That would be the case for `r` that you posted above.

**Comment by Amit-Modi** [comment]
Hi @certik, thanks for your help. I tried to do write a test, but it failed. Can you help me to point out my mistake.

**Comment by certik** [comment]
I think the error is unrelated. I restarted the failing test.

**Comment by gxyd** [comment]
Can you add a mention a few reference for the definition of this? For like definition of intersection of Ray with Circle/Ellipse. Also for `Segment`.

**Comment by gxyd** [comment]
Can you add a mention a few reference for the definition of this? For like definition of intersection of Ray with Circle/Ellipse. Also for `Segment`.
Add them in the file as well.

[Edit]: I mistakenly deleted the last comment.

**Comment by Amit-Modi** [comment]
@gxyd, sir the following definitions are taken from [http://jwilson.coe.uga.edu/emt669/student.folders/banker.teresa/unitstdy/unitstdy.html](http://jwilson.coe.uga.edu/emt669/student.folders/banker.teresa/unitstdy/unitstdy.html)
Definition: A tangent to a circle is a line in the plane of the circle that intersects the circle in exactly one point. The point of intersection is called the point of tangency.
unitstdy4
In the figure, line DC is a tangent; point B is the point of tangency. A ray or segment is tangent if it is a part of a tangent line and contains the point of tangency. So, ray BC is a tangent ray and segment BC is a tangent segment.

In which file I add this ?

**Comment by ylemkimon** [comment]
Please add it to `References` section like below in the docstring. See docstrings of other functions or classes for reference.
```
    References
    ==========
```

**Comment by Amit-Modi** [comment]
Thanks @ylemkimon, for your help. <br>And please tell me if anything else is required.

File: sympy/geometry/ellipse.py [hunk 1]
```diff
@@ -694,6 +694,11 @@ def is_tangent(self, o):
         >>> l1 = Line(p1, p2)
         >>> e1.is_tangent(l1)
         True
+        
+        References
+        ==========
+
+        [1] http://jwilson.coe.uga.edu/emt669/student.folders/banker.teresa/unitstdy/unitstdy.html
```
Comment (jksuom) [inline sympy/geometry/ellipse.py]: This `student.folders` link is likely to disappear in a short while. Would it be possible to find a more stable reference?

Comment (Amit-Modi) [reply sympy/geometry/ellipse.py]: @ylemkimon, I tried to find a more stable reference.<br> I mentioned earlier, I did not find a definition for tangent ray/segment (other than this ). <br>In most of the places tangent is defined as a tangent line, but I found following definition at (https://en.wikipedia.org/wiki/Tangent_(disambiguation))[https://en.wikipedia.org/wiki/Tangent_(disambiguation)]<br><br>More generally, in geometry, two curves are said to be tangent when they intersect at a given point and have the same direction at that point

**Comment by jksuom** [comment]
It seems to me that Change 2 is correct but Change 1 may be too strict. Two curves have generally several points in common, and they may be tangent to each other at some, but at not all of them. Hence it is usually said that the curves are tangent *at a certain point* instead of just saying that they are tangent. It is also sometimes said that the curves are tangent to each other if they are tangent at one point or several points.

It is also possible that two curves cross each other at a point where they have a common tangent. For example, consider the curve `y = x**3` in the `x-y`-plane that is tangent to the `x`-axis at the origin. 

It appears hard to find an easily accessible reference for the concept of tangency. Perhaps none would be necessary.

**Comment by gxyd** [comment]
@Amit-Modi can you please address the requested changes?

**Comment by vishal-vardhan** [comment]
@certik @gxyd @ylemkimon @jksuom can i work on this issue?

**Comment by Amit-Modi** [comment]
moved to #15774

**Comment by czgdp1807** [comment]
@Amit-Modi Is this issue resolved? I don't think no PR other than #15774 (which is open) referred this issue. Can you please reopen this issue. Thanks.

File: sympy/geometry/tests/test_ellipse.py [hunk 2]
```diff
@@ -378,6 +379,7 @@ def test_is_tangent():
     assert e1.is_tangent(Polygon((-3, 0), (3, 0), (0, 5))) is False
     assert e1.is_tangent(Polygon((-3, 0), (0, -5), (3, 0), (0, 5))) is False
     assert e1.is_tangent(Polygon((-3, -5), (-3, 5), (3, 5), (3, -5))) is True
+    assert e1.is_tangent(Polygon((3, 5), (3, -5), (-3, -5), (-3, 5))) is True
```
Comment (smichr) [inline sympy/geometry/tests/test_ellipse.py:382]: this has been fixed in #24716

File: sympy/geometry/tests/test_ellipse.py [hunk 3]
```diff
@@ -362,6 +362,7 @@ def test_is_tangent():
     assert c1.is_tangent(Ray((-3, -2), (-15, -20))) is False
     assert c1.is_tangent(Ray((-3, -22), (15, 20))) is False
     assert c1.is_tangent(Ray((9, 20), (9, -20))) is True
+    assert c1.is_tangent(Ray((2, 5), (9, 5))) is True
```
Comment (smichr) [inline sympy/geometry/tests/test_ellipse.py:365]: this behavior is still the same at this time.

**Pull request closed without merging**